contracts/Distributor.sol

```diff
@@ -0,0 +1,148 @@
+    priceLoadPercentage = _priceLoadPercentage;
+  }
+
+  function buyCover(
+        address coveredContractAddress,
+        bytes4 coverCurrency,
+        uint[] memory coverDetails,
+        uint16 coverPeriod,
+        uint8 _v,
+        bytes32 _r,
+        bytes32 _s
+  )
+     public
+     payable
+  {
+    uint requiredValue = priceLoadPercentage.mul(coverDetails[1]).div(100).add(coverDetails[1]);
+    require(msg.value == requiredValue, "Incorrect value sent");
+
+    Pool1.Pool1 p1 = Pool1.Pool1(nxMaster.getLatestAddress("P1"));
+    p1.makeCoverBegin.value(coverDetails[1])(coveredContractAddress, coverCurrency, coverDetails, coverPeriod, _v, _r, _s);
+
+    mintToken(coverCurrency, coverDetails, coverPeriod);
+  }
+
+  function buyCoverUsingCA(
+        address coveredContractAddress,
+        bytes4 coverCurrency,
+        uint[] memory coverDetails,
+        uint16 coverPeriod,
+        uint8 _v,
+        bytes32 _r,
+        bytes32 _s
+  )
+    public
+  {
+    uint requiredValue = priceLoadPercentage.mul(coverDetails[1]).div(100).add(coverDetails[1]);
+    PoolData.PoolData pd = PoolData.PoolData(nxMaster.getLatestAddress("PD"));
+    IERC20.IERC20 erc20 = IERC20.IERC20(pd.getCurrencyAssetAddress(coverCurrency));
+    require(erc20.transferFrom(msg.sender, address(this), requiredValue), "Transfer failed");
+
+    address payable pool1Address = nxMaster.getLatestAddress("P1");
+    Pool1.Pool1 p1 = Pool1.Pool1(pool1Address);
+    erc20.approve(pool1Address, coverDetails[1]);
+    p1.makeCoverUsingCA(coveredContractAddress, coverCurrency, coverDetails, coverPeriod, _v, _r, _s);
+
+    mintToken(coverCurrency, coverDetails, coverPeriod);
+  }
+
+  function mintToken(
+    bytes4 coverCurrency,
+    uint[] memory coverDetails,
+    uint16 coverPeriod
+  )
+  internal
+  {
+    QuotationData.QuotationData quotationData = QuotationData.QuotationData(nxMaster.getLatestAddress("QD"));
+    // *assumes* the newly created claim is appended at the end of the list covers
+    uint coverId = quotationData.getCoverLength().sub(1);
+
+    uint256 nextTokenId = tokenIdCounter++;
+    uint expirationTimestamp = block.timestamp + CLAIM_VALIDITY_MAX_DAYS_OVER_COVER_PERIOD + coverPeriod * 1 days;
+    allTokenData[nextTokenId] = TokenData(expirationTimestamp, msg.sender, coverCurrency, coverDetails, coverId);
+    _mint(msg.sender, nextTokenId);
+  }
+
+  function submitClaim(
+    uint256 tokenId
+  )
+    public
+    payable
+  {
+    require(_isApprovedOrOwner(msg.sender, tokenId), "Not approved or owner");
+    require(allTokenData[tokenId].expirationTimestamp > block.timestamp, "Token is expired");
+    require(allTokenData[tokenId].coverCurrency == "ETH", "currency not ETH");
+    uint coverAmount = allTokenData[tokenId].coverDetails[1];
+    require(msg.value == CLAIM_DEPOSIT_PERCENTAGE.mul(coverAmount).div(100), "Deposit value is incorrect");
+
+    _submitClaim(tokenId);
+  }
+
+  function submitClaimUsingCA(
```

Comment:
same as above, we could combine them into one submitClaim function for easier integration with frontend